master/internal/rm/agent_resource_manager.go

```diff
@@ -95,45 +101,83 @@ func (a AgentResourceManager) IsReattachEnabledForRP(ctx actor.Messenger, rpName
 	return false
 }
 
+// CheckMaxSlotsExceeded checks if the job exceeded the maximum number of slots.
+func (a AgentResourceManager) CheckMaxSlotsExceeded(
+	ctx actor.Messenger, name string, slots int,
+) (bool, error) {
+	agents, err := a.GetAgents(ctx, &apiv1.GetAgentsRequest{})
+	if err != nil {
+		return false, fmt.Errorf("unable to query agents for resource pool %v", name)
+	}
+	maxSlots := 0
+	for _, agent := range agents.Agents {
+		for _, resPool := range agent.ResourcePools {
+			if resPool == name {
+				maxSlots += len(agent.Slots)
+			}
+		}
+	}
+	if slots > maxSlots {
+		return true, nil
+	}
+	return false, nil
+}
+
 // ResolveResourcePool fully resolves the resource pool name.
 func (a AgentResourceManager) ResolveResourcePool(
 	ctx actor.Messenger, name string, slots int, command bool,
-) (string, error) {
+) (ResolvedResourcePool, error) {
+	var rp ResolvedResourcePool
 	// If the resource pool isn't set, fill in the default at creation time.
 	if name == "" && slots == 0 {
 		req := sproto.GetDefaultAuxResourcePoolRequest{}
 		resp, err := a.GetDefaultAuxResourcePool(ctx, req)
 		if err != nil {
-			return "", fmt.Errorf("defaulting to aux pool: %w", err)
+			return rp, fmt.Errorf("defaulting to aux pool: %w", err)
 		}
-		return resp.PoolName, nil
+		rp.Name = resp.PoolName
+		return rp, nil
 	}
 
 	if name == "" && slots >= 0 {
 		req := sproto.GetDefaultComputeResourcePoolRequest{}
 		resp, err := a.GetDefaultComputeResourcePool(ctx, req)
 		if err != nil {
-			return "", fmt.Errorf("defaulting to compute pool: %w", err)
+			return rp, fmt.Errorf("defaulting to compute pool: %w", err)
+		}
+		currentMaxSlotsExceeded, err := a.CheckMaxSlotsExceeded(ctx, name, slots)
+		if err != nil {
+			return rp, fmt.Errorf("validating request for (%s, %d): %w", name, slots, err)
 		}
-		return resp.PoolName, nil
+		rp.Name = resp.PoolName
+		rp.CurrentMaxSlotsExceeded = currentMaxSlotsExceeded
+		return rp, nil
 	}
 
 	if err := a.ValidateResourcePool(ctx, name); err != nil {
-		return "", fmt.Errorf("validating pool: %w", err)
+		return rp, fmt.Errorf("validating pool: %w", err)
```

Comment:
can you make the respond a pointer/default object, so we don't return something that looks half usable?